units/host_termination.go

```diff
@@ -235,7 +235,7 @@ func (j *hostTerminationJob) Run(ctx context.Context) {
 	// set host as decommissioned in DB so no new task will be assigned
 	prevStatus := j.host.Status
 	if prevStatus != evergreen.HostTerminated {
-		if err = j.host.SetDecommissioned(ctx, evergreen.User, false, "host will be terminated shortly, preventing task dispatch"); err != nil {
+		if err = j.host.SetDecommissioned(ctx, evergreen.User, false, j.TerminateIfBusy, "host will be terminated shortly, preventing task dispatch"); err != nil {
```

Comment:
I feel like because decommissioning is now dependent on `TerminateIfBusy`, there's now a bug lurking here between decommissioning and fixing the running task - we should be calling `SetDecommissioned` (along with checking if it's currently running a task) _before_ we try [fixing its running task](https://github.com/evergreen-ci/evergreen/blob/8f65301db2c88bc808cf828d3306ad0aa25ba957/units/host_termination.go#L186-L234). The reason being that the host termination job and task dispatch can still race (a task can still be dispatched in between [when the task is fixed](https://github.com/evergreen-ci/evergreen/blob/8f65301db2c88bc808cf828d3306ad0aa25ba957/units/host_termination.go#L198) and when it's decommissioned (here), meaning that it could fail to decommission the host (which is contradictory to its goal of decommissioning to prevent task dispatch). This would cause the just-dispatched task to also system-fail.

I don't think we necessarily need to fix it right now, but can we at least file a follow-up bug ticket for this?

Reply:
Created https://jira.mongodb.org/browse/DEVPROD-5507

Might be a silly question, why would we expect the newly dispatched task to system fail and not succeed as normal?

Reply:
This is trying to handle a case where [right above this](https://github.com/evergreen-ci/evergreen/blob/3130614c4be5555c8fe33d5045e76c0d2a0c84d3/units/host_termination.go#L198), Evergreen cleared the host's current task using `ClearAndResetStrandedHostTask` (which does something similar to `MarkEnd` + some restarting logic). After its task is de-assigned, the host is now free to take another task, which is bad because it's about to be terminated.

One way to stop it from accepting new tasks is to decommission it (the meaning of decommission in Evergreen is to finish the current task that the host is assigned, but don't accept any new ones, and try to terminate soon after finishing the current task). Unfortunately, if the host gets assigned a new task in between the `ClearAndResetStrandedHostTask` and here, that task has just been assigned to a host that's about to be terminated, meaning it'll system fail. I'm particularly worried about this race because `ClearAndResetStrandedHostTask` is effectively doing `MarkEnd` to system fail its current task. Since we know that `MarkEnd` is still slow, that gives the host plenty of time to pick up a task before it's decommissioned.

Reply:
Ah, this makes much more sense. I thought decommissioning was our way of terminating the tasks not the other way around where we decommission tasks that are about to be terminated. Thank you!